objects/src/assets/mod.rs

```diff
@@ -0,0 +1,420 @@
+        }
+    }
+}
+
+// FUNGIBLE ASSET
+// ================================================================================================
+/// A fungible asset.
+///
+/// A fungible asset consists of a faucet ID of the faucet which issued the asset as well as the
+/// asset amount. Asset amount is guaranteed to be 2^63 - 1 or smaller.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct FungibleAsset {
+    faucet_id: AccountId,
+    amount: u64,
+}
+
+impl FungibleAsset {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+    /// Specifies a maximum amount value for fungible assets which can be at most a 63-bit value.
+    pub const MAX_AMOUNT: u64 = (1u64 << 63) - 1;
+
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a fungible asset instantiated with the provided faucet ID and amount.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The provided faucet ID is not for a fungible asset faucet.
+    /// - The provide amount is greater than 2^63 - 1.
+    pub fn new(faucet_id: AccountId, amount: u64) -> Result<Self, AssetError> {
+        if !faucet_id.is_fungible_faucet() {
+            return Err(AssetError::not_a_fungible_faucet_id(faucet_id));
+        }
+
+        if amount > Self::MAX_AMOUNT {
+            return Err(AssetError::amount_too_big(amount));
+        }
+
+        Ok(Self { faucet_id, amount })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Return ID of the faucet which issued this asset.
+    pub fn faucet_id(&self) -> AccountId {
+        self.faucet_id
+    }
+
+    /// Returns the amount of this asset.
+    pub fn amount(&self) -> u64 {
+        self.amount
+    }
+
+    // OPERATIONS
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds two fungible assets together and returns the result.
+    ///
+    /// # Errors
+    /// Returns an error if the assets were not issued by the same faucet.
+    #[allow(clippy::should_implement_trait)]
+    pub fn add(self, other: Self) -> Result<Self, AssetError> {
+        if self.faucet_id != other.faucet_id {}
+
+        let amount = self.amount + other.amount;
+        if amount > Self::MAX_AMOUNT {
+            return Err(AssetError::amount_too_big(amount));
+        }
+
+        Ok(Self {
+            faucet_id: self.faucet_id,
+            amount,
+        })
+    }
+
+    /// Subtracts the specified amount from this asset and returns the resulting asset.
+    ///
+    /// # Errors
+    /// Returns an error if this asset's amount is smaller than the requested amount.
+    pub fn sub(&mut self, amount: u64) -> Result<Self, AssetError> {
+        if amount > self.amount {
+            return Err(AssetError::AmountNotSufficient(self.amount, amount));
```

Comment:
nit: I find this error a bit confusing, because the insufficient amount is not the amount provided to the function, but the amount in the account. In my way of thinking, the problem is that the requested amount is too big. Maybe it would be clearer if we qualified with `AssetAmountNotSufficient`(or insufficient)

Reply:
Agreed! Renamed.